### PR TITLE
[clang] Correct behavior of `LLVM_UNREACHABLE_OPTIMIZE=OFF` for `Release` builds

### LINKED ISSUE #68237: Cannot build `obj.clangAST` when `LLVM_UNREACHABLE_OPTIMIZATIONS` are turned off
I'm trying to build clang in `Release` mode with `LLVM_UNREACHABLE_OPTIMIZATIONS` turned off; however, the file `AttrImpl.cpp` fails to compile. This happens in `release/17.x` branch and `main`.

```bash
cmake llvm \
    -DLLVM_ENABLE_PROJECTS=clang \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DCMAKE_ASM_COMPILER=clang \
    -DLLVM_UNREACHABLE_OPTIMIZE=OFF \
    -B build
cd build/
make obj.clangAST
```

This seems due to the fact that on `CMAKE_BUILD_TYPE=Release` + `LLVM_UNREACHABLE_OPTIMIZE=OFF` leads to https://github.com/llvm/llvm-project/blob/6cbf6f5d376a06d96bc1320e83d479749873c654/llvm/include/llvm/Support/ErrorHandling.h#L151-L154 branch which causes the tblgen'd output (`AttrImpl.inc`) to not compile. All other variants work [(Release + `LLVM_UNREACHABLE_OPTIMIZE=ON`), (Debug + `LLVM_UNREACHABLE_OPTIMIZE=[ON|OFF]`)]

This is bisected to `clang-tblgen` change https://github.com/llvm/llvm-project/commit/b6d4d51f8f5aab311df34c753b925760578729bd#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491bR2642-R2646

Tested with both host toolchains `clang 17.0.2` and `gcc 13.2.0`

For Rust specific issue: Rust defaults to `LLVM_UNREACHABLE_OPTIMIZE=OFF` for all LLVM builds and this causes a regression when `LLVM` is compiled in release mode. See https://github.com/rust-lang/rust/pull/109373

### PR DESCRIPTION
**Pull request by MasterAwesome**
# Codegen
### Before
```c++
AArch64SVEPcsAttr *AArch64SVEPcsAttr::CreateImplicit(ASTContext &Ctx, SourceRange Range, Spelling S) {
  AttributeCommonInfo I(Range, NoSemaHandlerAttribute, (
    S == GNU_aarch64_sve_pcs ? AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, GNU_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/} :
    S == CXX11_clang_aarch64_sve_pcs ? AttributeCommonInfo::Form{AttributeCommonInfo::AS_CXX11, CXX11_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/} :
    S == C23_clang_aarch64_sve_pcs ? AttributeCommonInfo::Form{AttributeCommonInfo::AS_C23, C23_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/} :
    (llvm_unreachable("Unknown attribute spelling!"),  AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, 0, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/})));
  return CreateImplicit(Ctx, I);
}
```

### After
```c++
AArch64SVEPcsAttr *AArch64SVEPcsAttr::CreateImplicit(ASTContext &Ctx, SourceRange Range, Spelling S) {
  AttributeCommonInfo I(Range, NoSemaHandlerAttribute, [&]() {
    switch (S) {
    case GNU_aarch64_sve_pcs:
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, GNU_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    case CXX11_clang_aarch64_sve_pcs:
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_CXX11, CXX11_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    case C23_clang_aarch64_sve_pcs:
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_C23, C23_clang_aarch64_sve_pcs, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    default:
      llvm_unreachable("Unknown attribute spelling!");
      return AttributeCommonInfo::Form{AttributeCommonInfo::AS_GNU, 0, false /*IsAlignas*/, false /*IsRegularKeywordAttribute*/};
    }
  }());
  return CreateImplicit(Ctx, I);
}
```

Fixes #68237

### CODE REVIEW

**Review by AaronBallman — COMMENTED** [review]
Oh wow, that bug has been around for quite a while. I like your solution to it!

The code LGTM, no tests required because there isn't really a decent way to test this from lit. However, it would be nice to add a release note to `clang/docs/ReleateNotes.rst` to tell users that the bug was fixed. I'd add the note to the `Bug Fixes in This Version` section in the document.

**Comment by erichkeane** [comment]
Agree with Aaron on all points!

**Comment by MasterAwesome** [comment]
@erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`.

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -343,6 +343,8 @@ Bug Fixes in This Version
 - Fix a crash when evaluating value-dependent structured binding
   variables at compile time.
   Fixes (`#67690 <https://github.com/llvm/llvm-project/issues/67690>`_)
+- Fixes a regression where ``LLVM_UNREACHABLE_OPTIMIZE=OFF`` cannot be used
```
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:346]: Can we specify this was a Clang17 regression?

**Review by AaronBallman — APPROVED** [review]
LGTM modulo Erich's comment on the release notes.

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (MasterAwesome) [reply clang/docs/ReleaseNotes.rst]: Done

**Comment by MasterAwesome** [comment]
@AaronBallman, @erichkeane; whenever ya'll are done with the review, can you merge this in? I don't think I have access. And about the backporting, do we need it for 17.x bugfix releases?

**Review by AaronBallman — APPROVED** [review]
LGTM

**Comment by AaronBallman** [comment]
> @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue

Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.

**Comment by MasterAwesome** [comment]
> > @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue
> 
> Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.

Thanks for doing it! I'll work on cherry-picking this to Rust's fork of llvm :)

**Comment by AaronBallman** [comment]
> > > @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue
> > 
> > 
> > Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.
> 
> Thanks for doing it! I'll work on cherry-picking this to Rust's fork of llvm :)

You're welcome, but it seems there's a problem with trying to cherry-pick this over to 17.x and the pick will be more involved because of merge conflicts with the release notes. Would you mind taking over the pick to 17.x? If not, that's okay! (I'm mostly asking because I'm pretty swamped with reviews and I don't want the backport to fall through the cracks by accident.)

**Comment by MasterAwesome** [comment]
> > > > @erichkeane @AaronBallman, I've added the release notes and rebased the change on `main`. Does this need to be backported to 17.x? 16.x doesn't have this issue
> > > 
> > > 
> > > Yeah, I think we should probably backport this to 17.x. We have instructions on how to kick that off here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but I can get the ball rolling by adding the magic comment to the issue.
> > 
> > Thanks for doing it! I'll work on cherry-picking this to Rust's fork of llvm :)
> 
> You're welcome, but it seems there's a problem with trying to cherry-pick this over to 17.x and the pick will be more involved because of merge conflicts with the release notes. Would you mind taking over the pick to 17.x? If not, that's okay! (I'm mostly asking because I'm pretty swamped with reviews and I don't want the backport to fall through the cracks by accident.)

I gotchu! The PR created by llvmbot is at https://github.com/llvm/llvm-project-release-prs/pull/726

**Pull request merged**